compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/RangeContainsLowering.kt

```diff
@@ -0,0 +1,444 @@
+import org.jetbrains.kotlin.ir.symbols.IrSymbol
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.deepCopyWithSymbols
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.util.OperatorNameConventions
+import org.jetbrains.kotlin.utils.addIfNotNull
+
+val rangeContainsLoweringPhase = makeIrFilePhase(
+    ::RangeContainsLowering,
+    name = "RangeContainsLowering",
+    description = "Optimizes calls to contains() for ClosedRanges"
+)
+
+/**
+ * This lowering pass optimizes calls to contains() (`in` operator) for ClosedRanges.
+ *
+ * For example, the expression `X in A..B` is transformed into `A <= X && X <= B`.
+ */
+class RangeContainsLowering(val context: CommonBackendContext) : BodyLoweringPass {
+    override fun lower(irBody: IrBody, container: IrDeclaration) {
+        val transformer = Transformer(context, container as IrSymbolOwner)
+        irBody.transformChildrenVoid(transformer)
+    }
+}
+
+private class Transformer(
+    val context: CommonBackendContext,
+    val container: IrSymbolOwner
+) : IrElementTransformerVoidWithContext() {
+
+    private val headerInfoBuilder = RangeHeaderInfoBuilder(context, this::getScopeOwnerSymbol)
+    fun getScopeOwnerSymbol() = currentScope?.scope?.scopeOwnerSymbol ?: container.symbol
+
+    val stdlibExtensionContainsCallMatcher = SimpleCalleeMatcher {
+        extensionReceiver { it != null && it.type.isSubtypeOfClass(context.ir.symbols.closedRange) }
+        fqName { it == FqName("kotlin.ranges.${OperatorNameConventions.CONTAINS}") }
+        parameterCount { it == 1 }
+    }
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        // The call to contains() in `5 in 0..10` looks like this in IR:
+        //
+        //   CALL 'public open fun contains (value: kotlin.Int): kotlin.Boolean [operator] declared in kotlin.ranges.IntRange' type=kotlin.Boolean origin=IN
+        //
+        // And when `!in` is used in `5 !in 0..10`:
+        //
+        //   CALL 'public final fun not (): kotlin.Boolean [operator] declared in kotlin.Boolean' type=kotlin.Boolean origin=NOT_IN
+        //     $this: CALL 'public open fun contains (value: kotlin.Int): kotlin.Boolean [operator] declared in kotlin.ranges.IntRange' type=kotlin.Boolean origin=NOT_IN
+        //
+        // We only want to lower the call to contains(); even in the `!in` case, the call to not() should be preserved).
+        val origin = expression.origin
+        if (origin == null || (origin != IrStatementOrigin.IN && origin != IrStatementOrigin.NOT_IN)) {
+            return super.visitCall(expression)  // The call is not an `in` expression.
+        }
+        if (origin == IrStatementOrigin.NOT_IN && expression.symbol == context.irBuiltIns.booleanNotSymbol) {
+            return super.visitCall(expression)
+        }
```

Comment:
I don't quite understand why can't these be replaced with just
```
        if (origin != IrStatementOrigin.IN) return super.visitCall(expression)
```
?

Reply:
Currently in PSI2IR, for `!in`, _both_ the `not()` and the `contains()` calls have `origin=NOT_IN`. In this case, we only want to transform the `contains()` call, and not the `not()` call.

Sorry, it's not very obvious in the IR dump in the preceding comment. I'll modify the comment and explicitly call out the `NOT_IN` origins in the `!in` case. I also noticed just now that the `origin == null` clause is superfluous, so I'll remove that too.